### DISCUSSION TITLE
How do I use css variables with css modules in Next.JS

### DISCUSSION THREAD

**Discussion by motdde** [vercel/next.js#12147] [Help] [answered]
How do I make other modules share similar values like colours? I have tried using composes, it only works when I declare styles for each element as a class. @value syntax isn't working at all.

```
.links {
    //styles
}

.links a {
  //composes doesn't work here
}
```

**Comment by leerob** [comment]
Are you using the correct CSS module setup? 

https://nextjs.org/docs/basic-features/built-in-css-support#adding-component-level-css

Could you share an example?

**Comment by AWaselnuk** [comment]
I just ran into this same issue. Here is some setup to reproduce:

`/components/grid/Grid.module.css`

```css
@value test: #ff0000;

.item {
  background-color: test;
}
```

This results in this error:

```
Failed to compile
./components/grid/Grid.module.css (./node_modules/css-loader/dist/cjs.js??ref--5-oneOf-2-1!./node_modules/next/dist/compiled/postcss-loader??__nextjs_postcss!./components/grid/Grid.module.css)
CssSyntaxError

Selector ":export" is not pure (pure selectors must contain at least one local class or id)
```

My expectation is that this should work as specified in [CSS Module documentation](https://github.com/css-modules/css-modules/blob/master/docs/values-variables.md)

My first guess is that Next.js does not ship with the plugin `postcss-modules-values` which the CSS modules docs mention are required.

**Reply by robertsiipola** [reply]
Have you found a way around this? The docs mentions that out of the box, CSS variables are not compiled by PostCSS in Next when using the default configurations, due to incompatibility issues with CSS variables and IE 11: https://nextjs.org/docs/advanced-features/customizing-postcss-config#default-behavior

They suggest using scss, but I'm a bit bummed out that CSS variables don't seem to work by default.

**Comment by mulholo** [comment]
I know this isn't strictly what you're after, but we shifted to style-components after running into CSS variable limitations with CSS modules. Very happy with the decision.

**Reply by Bechrissed** [reply]
Ha! I'm just moving the opposite way :)

**Reply by BillyG83** [reply]
yeah so confused here, we are facing render issue where our app "pops in" and trying to move away from Styled Components to a more SSR friendly method

**Comment by mayank1513** [comment, marked as answer]
You can define variables inside global css and use them in all modules as if they were defined there.

**Reply by michael-goss** [reply]
using the same approach here 👍

**Reply by slp7711** [reply]
Great. This works for me. Tanks you.

**Reply by byteab** [reply]
``` css
:root {
    --primary: "red"
}
```
My mistake was defining color code inside a String

**Reply by tlchatt** [reply]
Any idea on a good way to do this with dynamic props from an api fetch?

**Reply by byteab** [reply]
maybe you should use styledcomponents

**Reply by thanhtutzaw** [reply]
my mistake was importing from global.css

**Reply by ScorpiosCrux** [reply]
> maybe you should use styledcomponents

styled components + variables = nice

**Reply by BillyG83** [reply]
@TheEhsanSarshar and @ScorpiosCrux Can you offer some insights into using Styled? We are thinking to move away from it as (even with the Next JS / Babel fixes) our app kind of "pops in" content. I was hoping moving to the CSS method recommended by Next would address this. I'd love your feedback thanks

**Reply by byteab** [reply]
you need to enable SSR for styled component to work properly in Next.js

**Comment by stephenscaff** [comment]
You have 2 options when using css vars with css modules. 

1. Define your css vars on :root in a global css file (as mentioned).
2. Define your css vars on an existing locally scoped class inside a CSS module file.

For example 

```
// style.module.scss (or less or css)

.item {
  --element-pad: 1em;
  // styles
š

.item__title {
  padding: var(--element-pad);
}
```

CSS vars don't _have_ to be on `:root`. They can be placed directly on a class, which will work in a CSS module file. Granted, this means the vars are only locally available, however, you can define vars on a parent component's class and they'll be available in child components.

**Reply by sytpb** [reply]
2 it is not work  in my module.css

**Reply by edahmitchel** [reply]
Not working in mine as well

**Reply by Matt-Jensen** [reply]
Above approach works, however you do have to parse in the generated class name from the module:

```html
<style>
{`.${styles.container} { --element-pad: 1em; }`}
</style>
```

**Comment by VelislavKarastoychev** [comment]
My solution is the following:
1.Go to the styles folder and create a file roots.css
2. Put all your css variables.
3. Go to pages/_app.js and import the '../styles/roots.css' file
4. Now you can use the variables in all *.module.css files of the project.

**Reply by vladio71** [reply]
It applied only after several project rebooting, so I thought it doesn`t work.)

**Comment by calvincchong** [comment]
Is there any way to pass in CSS variables the same way that styled components might be able to do? I'm trying to query the window scroll and pass in the appropriate variables as the style changes. Essentially making a parallax or scroll based animation. 

I read that someone was using styled components for this. Is this still the preferred way? I am using tailwind to get around the CSS limitation I just described but darn do I hate the long text bogging up my code

**Reply by byteab** [reply]
I personally use this way to pass css variables 
`<div style={{ '--size': '12px' }} />`

**Reply by calvincchong** [reply]
Hey @TheEhsanSarshar Thanks for this! I didn't realize that you can pass variables through the style attribute! Thank you! I didn't want to install yet another CSS dependency especially because we'd need to add another Babel plugin in order to be able to use serverside rendering
